pkgs/build-support/replace-vars/default.nix

```
@@ -0,0 +1,77 @@
+{ lib, stdenvNoCC }:
+
+/**
+  `replaceVars` is a wrapper around the [bash function `substitute`](https://nixos.org/manual/nixpkgs/stable/#fun-substitute)
+  in the stdenv. It allows for terse replacement of names in the specified path, while checking
+  for common mistakes such as naming a substitution that does nothing or forgetting a substitution.
+
+  As with the [`--subst-var-by`](https://nixos.org/manual/nixpkgs/stable/#fun-substitute-subst-var-by)
+  flag, names are encoded as `@name@` in the provided file at the provided path.
+
+  Any unmatched variable names in the file at the provided path will cause a build failure.
+
+  Any remaining text that matches `@[A-Za-z_][0-9A-Za-z_'-]@` in the output after replacement
+  has occurred will cause a build failure.
+
+  Attribute arguments:
+
+  - `path`: A [string-like value](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.strings.isStringLike) of a [store path](https://nixos.org/manual/nix/stable/store/store-path.html#store-path).
+
+  - `attrs`: each entry in this set corresponds to a `--subst-var-by` entry in `substitute`.
+    See https://nixos.org/manual/nixpkgs/stable/#fun-substitute
+
+  Example:
+
+  ```nix
+  { replaceVars }:
+
+  replaceVars ./greeting.txt { world = "hello"; }
+  ```
+
+  See `../../test/replace-vars/default.nix` for tests of this function.
+*/
+path: attrs:
+
+let
+  # We use `--replace-fail` instead of `--subst-var-by` so that if the thing isn't there, we fail.
+  subst-var-by = name: value: [
+    "--replace-fail"
+    (lib.escapeShellArg "@${toString name}@")
+    (lib.escapeShellArg (toString value))
```

```suggestion
    (lib.escapeShellArg value)
```

`toString` is a notoriously odd function for converting values to strings (it just concats lists using `" "` without any escaping!), so I think it's better to only support [string-like](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.strings.isStringLike) values.

---

I'll take your advice.

---

> it just concats lists using `" "` without any escaping

What's worse, is that it converts absolute paths to strings without context, creating "references" to paths that may not exist on the target host, like `/home/user/src/foo`.
Good riddance!